storage/spanner.go

```diff
@@ -0,0 +1,984 @@
+		return false, e
+	}
+	rowMap, err := createRowMap(r, colDDL, cols)
+	if err != nil {
+		return false, err
+	}
+	if expr != nil {
+		for index := 0; index < len(expr.Field); index++ {
+			status := evaluateStatementFromRowMap(expr.Condition[index], expr.Field[index], rowMap)
+			tmp, ok := status.(bool)
+			if !ok || !tmp {
+				if v1, ok := expr.AddValues[expr.Field[index]]; ok {
+
+					tmp, ok := rowMap[expr.Field[index]].(float64)
+					if ok {
+						m[expr.Field[index]] = tmp + v1
+						err = checkInifinty(m[expr.Field[index]].(float64), expr)
+						if err != nil {
+							return false, err
+						}
+					}
+				} else {
+					delete(m, expr.Field[index])
+				}
+			} else {
+				if v1, ok := expr.AddValues[expr.Field[index]]; ok {
+					tmp, ok := m[expr.Field[index]].(float64)
+					if ok {
+						m[expr.Field[index]] = tmp + v1
+						err = checkInifinty(m[expr.Field[index]].(float64), expr)
+						if err != nil {
+							return false, err
+						}
+					}
+				}
+			}
+			delete(expr.AddValues, expr.Field[index])
+		}
+		for k, v := range expr.AddValues {
+			val, ok := rowMap[k].(float64)
+			if ok {
+				m[k] = val + v
+				err = checkInifinty(m[k].(float64), expr)
+				if err != nil {
+					return false, err
+				}
+
+			} else {
+				m[k] = v
+			}
+		}
+	}
+	for i := 0; i < len(e.Attributes); i++ {
+		e.ValueMap[e.Tokens[i]] = evaluateStatementFromRowMap(e.Attributes[i], e.Cols[i], rowMap)
+	}
+
+	status, err := utils.EvaluateExpression(e)
+	if err != nil {
+		return false, err
+	}
+	return status, nil
+}
+
+func evaluateStatementFromRowMap(conditionalExpression, colName string, rowMap map[string]interface{}) interface{} {
+	if strings.HasPrefix(conditionalExpression, "attribute_not_exists") || strings.HasPrefix(conditionalExpression, "if_not_exists") {
+		if len(rowMap) == 0 {
+			return true
+		}
+		_, ok := rowMap[colName]
+		if ok {
+			return false
+		}
+		return true
+	}
+	if strings.HasPrefix(conditionalExpression, "attribute_exists") || strings.HasPrefix(conditionalExpression, "if_exists") {
+		if len(rowMap) == 0 {
+			return false
+		}
+		_, ok := rowMap[colName]
+		if ok {
+			return true
+		}
+		return false
+	}
+	return rowMap[conditionalExpression]
+}
+
+func (s Storage) performPutOperation(ctx context.Context, t *spanner.ReadWriteTransaction, table string, m map[string]interface{}) error {
+	ddl := models.TableDDL[table]
+	for k, v := range m {
+		t, ok := ddl[k]
+		if t == "BYTES(MAX)" && ok {
+			ba, err := json.Marshal(v)
+			if err != nil {
+				return errors.New("ValidationException", err)
+			}
+			m[k] = ba
+		}
+	}
+
+	mutation := spanner.InsertOrUpdateMap(table, m)
+	mutations := []*spanner.Mutation{mutation}
+	err := t.BufferWrite(mutations)
+	if e := errors.AssignError(err); e != nil {
+		return e
+	}
+	return nil
+}
+
+// SpannerBatchPut - this insert or update data in batch
+func (s Storage) SpannerBatchPut(ctx context.Context, table string, m []map[string]interface{}) error {
```

Comment:
This function should be combined with `SpannerPut` so they share the same logic. Now there seems to be some code duplication, and it seems that they use slightly different strategies for actually writing the data.

Reply:
@rosspatil, what are your thoughts around this